add /proc/net/unix parsing

peonone opened the pull request:
Signed-off-by: madk <luc.mdk@gmail.com>

SuperQ commented on net_unix.go:
```diff
@@ -0,0 +1,225 @@
+	netUnixStateDisconnected = 4
+)
+
+// NetUnixType is the type of the type field
```
Minor style nit, all comments should be full sentences. End these with a `.`.

SuperQ commented on net_unix.go:
```diff
@@ -0,0 +1,225 @@
+		Rows: make([]*NetUnixLine, 0, 32),
+	}
+	scanner := bufio.NewScanner(f)
+	// omit the header line
```
Similar here, but the sentences need to start with a capitol letter as well.

SuperQ commented on net_unix.go:
```diff
@@ -0,0 +1,225 @@
+
+func (u NetUnix) parseKernelPtr(str string) string {
+	length := len(str)
+	if length > 1 && str[length-1] == ':' {
```
I would use `strings. HasSuffix(str, ":")`.

SuperQ commented on net_unix.go:
```diff
@@ -0,0 +1,225 @@
+	return nuLine, nil
+}
+
+func (u NetUnix) parseKernelPtr(str string) string {
```
Shouldn't this function also return an error if this field doesn't end with a `:`? According to `af_unix.c#L2815`, it should always.

SuperQ commented:
Looks great, mostly minor nits.

peonone commented:
Fixed those nits.

Also updated the comment related to `Inode` field, and updated the type of this field to `uint64`, because it's  `unsigned long` in the code which generates the proc file.

grobie commented on net_unix.go:
```diff
@@ -0,0 +1,244 @@
+}
+
+// NetUnix holds the data read from /proc/net/unix.
+type NetUnix struct {
```
What's the motivation of using a struct over a simple slice `type NetUinx []*NetUnixLine`?

peonone replied:
There are few methods defined on `NetUinx` which can be called during the parsing process.
With a simple slice, the `NetUnix` itself changes once a line get parsed.
I think this way is not good and might cause potential bugs in future if we add more logic to the method, for example, pass the `NetUnix` instance to another place, but it points to the partial result. With a struct, we have the unique `NetUnix` instance.

grobie commented on net_unix.go:
```diff
@@ -0,0 +1,244 @@
+}
+
+func (u *NetUnix) parseLine(line string, hasInode bool) (*NetUnixLine, error) {
+	fields := strings.Fields(line)
```
Please check the length of the fields first to prevent out of bounds panics. We've seen all kinds of unexpected errors in procfs, from malformed lines, to unexpected blank lines and unexpected format changes. The library should handle these gracefully.

grobie commented on net_unix.go:
```diff
@@ -0,0 +1,244 @@
+		Type:      typ,
+		Flags:     flags,
+		State:     state,
+		Path:      fields[len(fields)-1],
```
The path attribute can be optional, this needs to be handled.

grobie commented on net_unix.go:
```diff
@@ -0,0 +1,244 @@
+	}
+	var inode uint64
+	if hasInode {
+		inodeStr := fields[len(fields)-2]
```
This doesn't work on lines without a path. Count from the beginning.

peonone commented:
@grobie 's comments make sense, I added validation for the length and handled the case of a blank path.

mdlayher commented on net_unix.go:
```diff
@@ -0,0 +1,268 @@
+
+// NewNetUnixByPath returns data read from /proc/net/unix by file path.
+// It might returns an error with partial parsed data, if an error occur after some data parsed.
+func NewNetUnixByPath(path string) (*NetUnix, error) {
```
Nit: I would accept `r io.Reader` in this function and open a file in the method above instead.  This gives you and your caller the most flexibility.

mdlayher commented on net_unix.go:
```diff
@@ -0,0 +1,268 @@
+	// Omit the header line.
+	scanner.Scan()
+	header := scanner.Text()
+	// From the man page of proc(5), it dose not contain an Inode field,
```
*does

mdlayher commented on net_unix.go:
```diff
@@ -0,0 +1,268 @@
+	return str[:len(str)-1], nil
+}
+
+func (u NetUnix) parseUsers(hexStr string) (uint64, error) {
```
Nit: it feels a little weird to have these be methods when they don't actually do anything with the method receiver at all, but I suppose it makes sense to avoid littering the package with unexported functions that are only used for a very specific file.

peonone replied:
Yes, the purpose of making them as methods is preventing package level functions with a prefix.

mdlayher commented on net_unix.go:
```diff
@@ -0,0 +1,268 @@
+func (f NetUnixFlags) String() string {
+	switch f {
+	case netUnixFlagListen:
+		return "accepton"
```
Nit: should this be "listen" to match the flag name?  I'm not familiar with this file.

peonone replied:
I referred to the [source code](https://elixir.bootlin.com/linux/v4.17/source/net/unix/af_unix.c#L2819) which writes the proc file, `listen` also makes sense and easier to understand, I will update it.

pgier commented:
@peonone any chance you can rebase this one against latest master?  Sorry this wasn't merged earlier.  After fixing the conflict I think we can merge it.

peonone commented:
> @peonone any chance you can rebase this one against latest master? Sorry this wasn't merged earlier. After fixing the conflict I think we can merge it.

@pgier Sure, I will do this later this week.

peonone commented:
@pgier Done
